### PR TITLE
[MERGE] website: use new OWL components for page properties

### PR DESCRIPTION
**Pull request by xO-Tx**
Follows the merge of the 'website in backend' task at [1] and the new
OWL views structure at [2].

See sub-commits for details.

[1]: https://github.com/odoo/odoo/commit/31cc10b91dc7762e23b4bde9b945be0c4ce3fe3b
[2]: https://github.com/odoo/odoo/commit/48ef812a635f70571b395f82ffdb2969ce99da9e

task-2687506

### CODE REVIEW

File: addons/website/static/src/components/dialog/page_properties.js [hunk 1]
```diff
@@ -155,108 +154,122 @@ DuplicatePageDialog.template = xml`
...
 DuplicatePageDialog.props = {
-    onClose: Function,
-    close: Function,
+    onDuplicate: {type: Function, optional: true},
     pageId: Number,
+    close: Function,
+};
+
+export const pagePropertiesProps = {
+    title: _t('Page Properties'),
```
Comment (qsm-odoo) [inline addons/website/static/src/components/dialog/page_properties.js:163]: `"` instead of `'` for translatable strings

File: addons/website/static/src/components/dialog/page_properties.js [hunk 2]
```diff
@@ -155,108 +154,122 @@ DuplicatePageDialog.template = xml`
 </WebsiteDialog>
 `;
 DuplicatePageDialog.props = {
-    onClose: Function,
-    close: Function,
+    onDuplicate: {type: Function, optional: true},
     pageId: Number,
+    close: Function,
```
Comment (qsm-odoo) [inline addons/website/static/src/components/dialog/page_properties.js:159]: I don't understand why this line is moved by your commit? If it was to order by alphabetical order maybe but if it is for no reason at all, avoid creating useless diff :+1:

File: addons/website/static/src/components/dialog/page_properties.js [hunk 3]
```diff
@@ -155,108 +154,122 @@ DuplicatePageDialog.template = xml`
...
-            this.orm = useService('orm');
-        } catch {
-            // Use services in legacy environment.
-            this.websiteService = useWowlService('website');
-            this.dialogService = useWowlService('dialog');
-            this.orm = useWowlService('orm');
-        }
+        super.setup();
+        this.dialog = useService('dialog');
+        this.website = useService('website');
```
Comment (qsm-odoo) [inline addons/website/static/src/components/dialog/page_properties.js:175]: Mmh... I see that half the time we used "Service" in the name and half the time we did not :grimacing: I see that the framework JS team seems to use the version without the service word indeed. If you confirm it, could you make a dedicated commit which renames those variables in all of our components first? So that we avoid diff in these commits and so that we enforce the convention everywhere once and for all :+1: (either in another PR or this one, as you prefer. If there is many things to review after my review, maybe do that in another PR first).

File: addons/website/static/src/components/dialog/page_properties.js [hunk 4]
```diff
@@ -155,108 +154,122 @@ DuplicatePageDialog.template = xml`
...
+            this.actionClone = this.modalRef.el.querySelector('.o_page_properties_clone');
+            this.actionDelete = this.modalRef.el.querySelector('.o_page_properties_delete');
+            this.actionClone.addEventListener('click', this.clonePage.bind(this));
+            this.actionDelete.addEventListener('click', this.deletePage.bind(this));
+        });
+
+        onWillUnmount(() => {
+            this.actionClone.removeEventListener('click', this.clonePage.bind(this));
+            this.actionDelete.removeEventListener('click', this.deletePage.bind(this));
+        });
```
Comment (qsm-odoo) [inline addons/website/static/src/components/dialog/page_properties.js:190]: Should not this be done via a `t-on-click` in the OWL template? And would mean 0-2 XML lines instead of 11 JS lines? (real question, I am a OWL noob)

File: addons/website/static/src/components/dialog/page_properties.xml [hunk 5]
```diff
@@ -8,14 +8,14 @@
...
-                <div t-att-data-template="depTemplate" data-html="true" title="Dependencies" t-ref="action">
+                <div t-att-data-bs-template="depTemplate" data-bs-html="true" title="Dependencies" t-ref="action">
                     It looks like your file is being called by
                     <a href="#" t-on-click="showDependencies"><t t-esc="depText" /></a>.
                     Changing its name will break these calls.
                 </div>
             </t>
             <t t-else="">
-                <div t-att-data-template="depTemplate" data-html="true" title="Dependencies" t-ref="action">
+                <div t-att-data-bs-template="depTemplate" data-bs-html="true" title="Dependencies" t-ref="action">
```
Comment (qsm-odoo) [inline addons/website/static/src/components/dialog/page_properties.xml:18]: Should this be a dedicated commit? I guess it is a BS5 fix, not an OWL conversion?

File: addons/web/static/src/views/view_dialogs/form_view_dialog.js [hunk 6]
```diff
@@ -115,6 +115,7 @@ FormViewDialog.props = {
     preventCreate: { type: Boolean, optional: true },
     preventEdit: { type: Boolean, optional: true },
     isToMany: { type: Boolean, optional: true },
+    size: { type: String, optional: true },
```
Comment (qsm-odoo) [inline addons/web/static/src/views/view_dialogs/form_view_dialog.js:118]: I see that the `size` props of the Dialog component is defined with`size: { type: String, optional: true, validate: (s) => ["sm", "md", "lg", "xl"].includes(s) },` and also sets a default to `lg` via `defaultProps`. Should not we do the same here? Even if I guess the validation is done in the end by the Dialog sub component but making it sooner might be better (although I guess it would be done twice...)? But mainly it would prevent code duplication and make it clear that this is the Dialog props you allow the control here. Something like: `size: Dialog.props.size`. Maybe confirm with the framework team if needed.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Ping me once you are finished adapting so I'll merge this work in priority :+1:

File: addons/website/static/src/components/dialog/page_properties.js (hunk 2 above)
Comment (xO-Tx) [reply addons/website/static/src/components/dialog/page_properties.js]: I checked the DOC on Owl repo & /web code and there is nothing that confirmed the order of props... so I just restored the diff :+1:

File: addons/website/static/src/components/dialog/page_properties.js (hunk 3 above)
Comment (xO-Tx) [reply addons/website/static/src/components/dialog/page_properties.js:166]: > I see that the framework JS team seems to use the version without the service word indeed

I actually did this because they use it without "Service" most of the time, but I don't think there is a convention about this (on [Odoo frontend DOC](https://www.odoo.com/documentation/master/developer/reference/frontend/services.html?highlight=useservice) I found both (with & without "Service"), so I'm not sure about this...?

File: addons/website/static/src/components/dialog/page_properties.js (hunk 4 above)
Comment (xO-Tx) [reply addons/website/static/src/components/dialog/page_properties.js:181]: Actually, this is a special case:
The `FormViewDialog` component (which is extended by `PagePropertiesDialog`) has the `View` as sub-component, so the `buttonTemplate` prop (footer actions template) is passed to the view to override the original one used by formViews (`t-name="web.FormView.Buttons"` [A]), so to have the new page properties buttons (Duplicate/Delete page), I had to pass the new XML in `buttonTemplate`, which means I can't use `t-on-click` simply because I can't handle them outside of `PagePropertiesDialog`, ([A] is linked to the formView), so I had to handle them directly from the DOM, I discussed this With ARD last time and he suggested to use the `useListener` hook (which does the same thing as the current code in a hook) but after checking the code, it seems [deprecated](https://github.com/odoo/odoo/blob/8a5721bc7d5540e382d2d06fe5b9634911b7859c/addons/web/static/src/core/utils/hooks.js#L86) since it uses the old `component.el` attribute (details [here](https://github.com/odoo/odoo/commit/8a5721bc7d5540e382d2d06fe5b9634911b7859c)).

File: addons/website/static/src/components/dialog/page_properties.xml (hunk 5 above)
Comment (xO-Tx) [reply addons/website/static/src/components/dialog/page_properties.xml:18]: Added in a separate commit :+1:

File: addons/website/static/src/components/fields/fields.js [hunk 7]
```diff
@@ -1,24 +1,32 @@
...
 
-const {useState, onWillStart} = owl;
+const {Component, useState, onWillStart} = owl;
 
-class FieldPageUrl extends AbstractFieldOwl {
+/**
+ * Displays website page dependencies and URL redirect options when the page URL
+ * is updated.
+ */
+class PageUrlField extends Component {
```
[the remark this answers is not in the document]
Comment (xO-Tx) [reply addons/website/static/src/components/fields/fields.js:19]: Yes we don't need to use the `AbstractField`. Actually we had a different structure in the old version:

A- Owl specific fields should extend: [`AbstractField` which extends `LegacyComponent`(*)], which is just a translation of `AbstractField` widget features to Owl with an override on `BasicRenderer` to handle both legacy & Owl fields on legcay views (details [here](https://github.com/odoo/odoo/commit/3fd7b2009ec1f16c56c6d0ae87004011c90e204c)).

B- But now since the [new Owl views structure](https://github.com/odoo/odoo/commit/48ef812a635f70571b395f82ffdb2969ce99da9e):

Fields can be added as normal Components with some updates to make the new views API work (e.g. we add related field props (see: `standardFieldProps`), the `extractProps` API used to extract data from the field in view arch instead of having them ready on widget in legacy code...)

I didn't found any documentation for the new structure but I think you can see in /web > `FormController`, `FormRenderer`, `View` and `Field` components how this was implemented :+1: 

(*): LegacyComponent is depercated now.

File: addons/web/static/src/views/view_dialogs/form_view_dialog.js (hunk 6 above)
Comment (xO-Tx) [reply addons/web/static/src/views/view_dialogs/form_view_dialog.js]: Well, since the `PagePropertiesDialog` is added using dialog service (using a `DialogContainer` component which contains dialogs as sub-components with their dynamic props...), I'm not sure the "DEV mode props validation" will be done in this case (to be confirmed by @odoo/rd-framework-js) but I think the `size: Dialog.props.size description` description is better anyway :+1:

File: addons/website/static/src/components/dialog/page_properties.js (hunk 3 above)
Comment (qsm-odoo) [reply addons/website/static/src/components/dialog/page_properties.js:166]: Mmh, they seem to do both in the doc indeed... ok, let's not waste time and keep your diff about that in this commit. I guess we'll enforce a convention over time...

File: addons/website/static/src/components/dialog/page_properties.js (hunk 4 above)
Comment (qsm-odoo) [reply addons/website/static/src/components/dialog/page_properties.js:181]: I see. I guess you can't use t`-on-click` directly but what you do here seems a bit breaking the OWL logic IMO: you have a component (a "dialog") which contains another (a "view") but the "dialog", once mounted, search the "view" DOM for stuff and add event listeners. Should not you rather give props (from the dialog to the view) which are the handlers to call for the buttons? So in your button template, something like `t-on-click="props.clonePage"` (via `viewProps` that I see ?). If not that, a variant of it. I am sure there is a way for the button section to "communicate" with the dialog component (either via props that are given by the dialog as I explained or another system) but I doubt the framework team would approve of searching the DOM of a subcomponent... :grimacing: In doubt, check with them, they don't bite (I think) :wink: (cc @detrouxdev as you were mentioned)

Comment (xO-Tx) [reply addons/website/static/src/components/dialog/page_properties.js]: I updated the code to override the `FormViewDialog` related template and add buttons directly in the footer's slot (I just forgot that I have this possibility) Thanks :+1:

**Review by qsm-odoo — APPROVED** [review]
The diff and split into commits seems good @xO-Tx but the commits were not working on their own (if I take the ForwViewDialog commit only, it seems to work but fields do not behave the correct way). I push-forced this PR to keep the two commits related to OWL conversion and make a merge commit linking them together. I moved the other two commits to this PR which I'll merge later: https://github.com/odoo/odoo/pull/98180

I do think there are a few things to improve in this work but not worth slowing down our dev, this is a good first step :+1: Further refactoring will follow :+1: 

@robodoo rebase-merge r+
